File: leetcode/3sum.py
# 时间复杂度 O(n^2)
# 空间复杂度 O(1)
class Solution:
    def threeSum(self, nums: [int], target) -> [[int]]:
        n = len(nums)
        if n < 3:
            return []
        result = []
        nums.sort()
        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            kIndex = n - 1
            for j in range(i + 1, n):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                while j < kIndex and nums[i] + nums[j] + nums[kIndex] > target:
                    kIndex = kIndex - 1
                if j == kIndex:
                    break
                if nums[i] + nums[j] + nums[kIndex] == target:
                    temp = [nums[i], nums[j], nums[kIndex]]
                    if temp not in result:
                        result.append(temp)
        return result


# 时间复杂度 O(n^2)
# 空间复杂度 O(n)
class Solution2:
    def threeSum(self, nums: [int], target) -> [[int]]:
        n = len(nums)
        if n < 3:
            return []
        result = []
        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            map = {}
            for j in range(i + 1, n):
                t = target - nums[i] - nums[j]
                if t in map and map[t] != j:
                    temp = [nums[i], nums[j], t]
                    temp.sort()
                    if temp not in result:
                        result.append(temp)
                else:
                    map[nums[j]] = j
        return result


print(Solution2().threeSum([-1, 0, 1, 2, -1, -4], 2))

File: leetcode/3sum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            List<Integer> temp = twoSum(nums, nums[i], i);
            res.add(temp);
        }
        return res;
    }

    public List<Integer> twoSum (int[] numbers, int target, int skipIndex) {
        HashMap<Integer, Integer> map = new HashMap<>();
        List<Integer> res = new ArrayList<>();

        for (int i = 0; i < numbers.length; i++) {
            if (i == skipIndex) {
                continue;
            }
            int temp = target - numbers[i];
            if (map.containsKey(numbers[i])) {
                res.add(target);
                res.add(numbers[i]);
                res.add(numbers[skipIndex]);
                Collections.sort(res);
                return res;
            } else {
                map.put(temp, i);
            }
        }
        return null;
    }
}